pynumdiff/kalman_smooth/_kalman_smooth.py

```diff
@@ -1,643 +1,259 @@
+    param float r: variance of the signal noise
+    param float q: variance of the constant velocity model
+    :param bool forwardbackward: indicates whether to run smoother forwards and backwards
+        (usually achieves better estimate at end points)
+
+    :return: tuple[np.array, np.array] of\n
+        - **x_hat** -- estimated (smoothed) x
+        - **dxdt_hat** -- estimated derivative of x
     """
-    if options is None:
-        options = {'backward': False}
-
-    r, q = params
-
-    if len(x.shape) == 2:
-        y = x
-    else:
-        y = np.reshape(x, [1, len(x)])
-
-    A = np.array([[1, dt], [0, 1]])
-    B = np.array([[0], [0]])
-    C = np.array([[1, 0]])
+    if params != None: # boilerplate backwards compatibility code
+        warn("""`params` and `options` parameters will be removed in a future version. Use `r`,
+            `q`, and `forwardbackward` instead.""", DeprecationWarning)
+        r, q = params
+        if options != None:
+            if 'forwardbackward' in options: forwardbackward = options['forwardbackward']
+    elif r == None or q == None:
+        raise ValueError("`q` and `r` must be given.")
+
+    A = np.array([[1, dt], [0, 1]]) # states are x, x'. This basically does an Euler update.
+    C = np.array([[1, 0]]) # we measure only y = noisy x
     R = np.array([[r]])
-    Q = np.array([[1e-16, 0], [0, q]])
-    x0 = np.array([[x[0,0]], [0]])
-    P0 = np.array(100*np.eye(2))
-    u = None
-
-    
-
-    if options['backward']:
-        A = np.linalg.pinv(A)
-        y = y[:, ::-1]
-
-    xhat_fp, xhat_fm, P_fp, P_fm = __kalman_forward_filter__(x0, P0, y, u, A, B, C, R, Q)
-    xhat_smooth, _ = __kalman_backward_smooth__(xhat_fp, xhat_fm, P_fp, P_fm, A)
-
-    x_hat = np.ravel(xhat_smooth[0, :])
-    dxdt_hat = np.ravel(xhat_smooth[1, :])
-
-    if not options['backward']:
-        return x_hat, dxdt_hat
-
-    return x_hat[::-1], dxdt_hat[::-1]
-
-
-def constant_velocity(x, dt, params, options=None):
-    """
-    Run a forward-backward constant velocity RTS Kalman smoother to estimate the derivative.
-
-    :param x: array of time series to differentiate
-    :type x: np.array (float)
-
-    :param dt: time step size
-    :type dt: float
-
-    :param params: a list of two elements:
-
-                    - r: covariance of the x noise
-                    - q: covariance of the constant velocity model
-
-    :type params: list (float)
-
-
-    :param options: a dictionary indicating whether to run smoother forwards and backwards
-                    (usually achieves better estimate at end points)
-    :type params: dict {'forwardbackward': boolean}, optional
-
-    :return: a tuple consisting of:
-
-            - x_hat: estimated (smoothed) x
-            - dxdt_hat: estimated derivative of x
-
-
-    :rtype: tuple -> (np.array, np.array)
-    """
-    if len(x.shape) == 2:
-        pass
-    else:
-        x = np.reshape(x, [1, len(x)])
```

Comment:
It's not necessary to reshape 1D vectors into vertical 2D column vectors, because `A @ x` will do the exact same thing in either case.